kafka-client/src/main/java/dev/responsive/kafka/clients/ResponsiveProducer.java

```diff
@@ -144,46 +90,109 @@ public void sendOffsetsToTransaction(
       final Map<TopicPartition, OffsetAndMetadata> offsets,
       final String consumerGroupId
   ) throws ProducerFencedException {
-    offsets.forEach((p, o) -> offsetRecorder.record(consumerGroupId, p, o.offset()));
     wrapped.sendOffsetsToTransaction(offsets, consumerGroupId);
+    for (final var l : listeners) {
+      l.onSendOffsetsToTransaction(offsets, consumerGroupId);
+    }
   }
 
   @Override
   public void sendOffsetsToTransaction(
       final Map<TopicPartition, OffsetAndMetadata> offsets,
       final ConsumerGroupMetadata groupMetadata
   ) throws ProducerFencedException {
-    offsets.forEach((p, o) -> offsetRecorder.record(groupMetadata.groupId(), p, o.offset()));
     wrapped.sendOffsetsToTransaction(offsets, groupMetadata);
+    for (final var l : listeners) {
+      l.onSendOffsetsToTransaction(offsets, groupMetadata.groupId());
+    }
   }
 
   @Override
   public void commitTransaction() throws ProducerFencedException {
     wrapped.commitTransaction();
-    final Map<RecordingKey, Long> offsets = offsetRecorder.onCommit();
-    for (final var listener : listeners) {
-      try {
-        listener.onCommit(offsets);
-      } catch (final Throwable t) {
-        logger.error("error from responsive producer commit listener", t);
-      }
-    }
+    listeners.forEach(Listener::onCommit);
   }
 
   @Override
   public void abortTransaction() throws ProducerFencedException {
     wrapped.abortTransaction();
-    offsetRecorder.onAbort();
+    listeners.forEach(Listener::onAbort);
+  }
+
+  private static class RecordingFuture implements Future<RecordMetadata> {
```

Comment:
suggestion: can we move inner classes to the bottom of the file? it makes it a bit easier to read through the flow